feat: check detector consistency

niermann999 opened the pull request:
Add a number of utility functions that perform some simple consistency checks on a detector object 
and its volumes and surfaces:

  -  check for empty data stores
  -  check for any invalid links in volumes and surfaces
  -  check that volumes and surfaces have the correct volume index set
  -  check that mask boundaries are consistent
  -  check that sensitive/passive and portals surfaces have a navigation link either to their mother volume or not, respectively
  -  check that every surface in the brute force method/grids can be found via the detector lookup and vice versa

To avoid a segfault during the checking run, I added an empty (and in the IO a dummy) volume grid to the detector. 
This will be handled properly in a future PR.

To check different detectors read in from json, I added a folder called checks, where 
all of the consistency check will be called in the future. There is a test now that 
inherits from the google test test via a new default test fixture, which call the new
consistency check.

beomki-yeo commented:
OK

beomki-yeo commented on core/include/detray/core/detector.hpp:
```diff
@@ -502,6 +502,14 @@ class detector {
+    ///
+    /// @param v_grid the volume grid to be added
+    DETRAY_HOST
+    inline auto add_volume_finder(const volume_finder &v_grid) -> void {
```
Shouldn't it be `set_volume_finder` instead?

niermann999 replied:
Renamed

beomki-yeo commented on core/include/detray/masks/line.hpp:
```diff
@@ -150,14 +152,42 @@ class line {
+    DETRAY_HOST constexpr bool check_consistency(
+        const bounds_t<scalar_t, kDIM> &bounds, std::ostream &os) const {
+
+        constexpr auto epsilon{10.f * std::numeric_limits<scalar_t>::epsilon()};
```
Why do you multiply 10 here?

niermann999 replied:
Because ```std::numeric_limits<scalar_t>::epsilon()``` is the minimal difference between two floating point numbers, which I thought is too small for a tolerance on mask values. Let me rename the tolerance

beomki-yeo reviewed (commented):
I like the PR but can we have consistency checker in a more general place instead of `test` directory? (e.g. detray/utils) It would be easier to use it in other projects

niermann999 commented:
> I like the PR but can we have consistency checker in a more general place instead of `test` directory? (e.g. detray/utils) It would be easier to use it in other projects

Absolutely. See #522 :)

niermann999 commented:
Update after rebase: Using the detector reader in consistency check executable now, moved the volume grid building to the ```detector_builder``` and moved some type traits around, so that the grid builder does not need to be included everywhere

beomki-yeo commented:
Looks good
